pkg/imgutil/dockerconfigresolver/credentialsstore_test.go

```diff
@@ -0,0 +1,384 @@
+				if err != nil {
+					t.Fatal(err)
+				}
+				return tmpDir
+			},
+			errorNew: ErrUnableToInstantiate,
+		},
+		// FIXME: this test works in Lima, but does not work on github action
+		// It is unclear why at this point.
+		/*
+			{
+				description: "Pointing DOCKER_CONFIG at a directory containing a `config.json` dangling symlink will prevent saving credentials",
+				setup: func() string {
+					tmpDir := createTempDir(t, 0700)
+					err := os.Symlink("doesnotexist", filepath.Join(tmpDir, "config.json"))
+					if err != nil {
+						t.Fatal(err)
+					}
+					return tmpDir
+				},
+				errorWrite: ErrUnableToStore,
+			},
+		*/
+		{
+			description: "Pointing DOCKER_CONFIG at a directory containing an unreadable, valid `config.json` file will prevent instantiation",
+			setup: func() string {
+				tmpDir := createTempDir(t, 0700)
+				err := os.WriteFile(filepath.Join(tmpDir, "config.json"), []byte("{}"), 0600)
+				if err != nil {
+					t.Fatal(err)
+				}
+				err = os.Chmod(filepath.Join(tmpDir, "config.json"), 0000)
+				if err != nil {
+					t.Fatal(err)
+				}
+				return tmpDir
+			},
+			errorNew: ErrUnableToInstantiate,
+		},
+		{
+			description: "Pointing DOCKER_CONFIG at a directory containing a read-only, valid `config.json` file will NOT prevent saving credentials",
+			setup: func() string {
+				tmpDir := createTempDir(t, 0700)
+				err := os.WriteFile(filepath.Join(tmpDir, "config.json"), []byte("{}"), 0600)
+				if err != nil {
+					t.Fatal(err)
+				}
+				err = os.Chmod(filepath.Join(tmpDir, "config.json"), 0400)
+				if err != nil {
+					t.Fatal(err)
+				}
+				return tmpDir
+			},
+		},
+	}
+
+	t.Run("Broken Docker Config testing", func(t *testing.T) {
+		registryURL, err := Parse("registry")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		for _, tc := range testCases {
+			t.Run(tc.description, func(t *testing.T) {
+				directory := tc.setup()
+				cs, err := NewCredentialsStore(directory)
+				assert.ErrorIs(t, err, tc.errorNew)
+				if err != nil {
+					return
+				}
+
+				var af *Credentials
+				af, err = cs.Retrieve(registryURL, true)
+				assert.ErrorIs(t, err, tc.errorRead)
+
+				err = cs.Store(registryURL, af)
+				assert.ErrorIs(t, err, tc.errorWrite)
+			})
+		}
+	})
+}
+
+func writeContent(t *testing.T, content string) string {
+	t.Helper()
+	tmpDir := createTempDir(t, 0700)
+	err := os.WriteFile(filepath.Join(tmpDir, "config.json"), []byte(content), 0600)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return tmpDir
+}
+
+func TestWorkingCredentialsStore(t *testing.T) {
+	testCases := []struct {
+		description string
+		setup       func() string
+		username    string
+		password    string
+	}{
+		{
+			description: "Reading credentials from `auth` using canonical identifier",
+			username:    "username",
+			password:    "password",
+			setup: func() string {
+				content := `{
+				"auths": {
+					"someregistry:443": {
+						"auth": "dXNlcm5hbWU6cGFzc3dvcmQ="
+					}
```

Comment:
Maybe this should be directly encoded in the test?
I wonder if some credential leakage detection tool may report a false positive for this.

Reply:
Interesting...